Source/Provers/LeanAuto/LeanAuto.cs

```diff
@@ -0,0 +1,1035 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace Microsoft.Boogie.LeanAuto;
+
+internal class LeanConversionException : Exception
+{
+  public string Msg { get; }
+  public LeanConversionException(IToken tok, string s)
+  {
+    Msg = tok + ": " + s;
+  }
+}
+
+public class LeanGenerator : ReadOnlyVisitor
+{
+  private readonly TextWriter writer;
+  private readonly VCGenOptions options;
+  private readonly List<NamedDeclaration> globalVars = new();
+  private readonly HashSet<string> usedNames = new();
+  private bool usesMaps;
+  private readonly List<string> mapAxiomNames =
+    new(new[]
+    {
+      "SelectStoreSame1", "SelectStoreDistinct1",
+      "SelectStoreSame2", "SelectStoreDistinct2"
+    });
+  private readonly List<string> userAxiomNames = new();
+  private readonly Dictionary<Type, HashSet<string>> uniqueConsts = new();
+
+  private const string AndString = "\u2227"; /* ∧ */
+  private const string OrString = "\u2228"; /* ∨ */
+  private const string NeqString = "\u2260"; /* ≠ */
+  private const string ArrowString = "\u2192"; /* → */
+
+  private LeanGenerator(VCGenOptions options, TextWriter writer)
+  {
+    this.options = options;
+    this.writer = writer;
+  }
+
+  public static void EmitPassiveProgramAsLean(VCGenOptions options, Program p, TextWriter writer)
+  {
+    var generator = new LeanGenerator(options, writer);
+    generator.EmitHeader();
+    try {
+      var allBlocks = p.Implementations.SelectMany(i => i.Blocks);
+      var liveDeclarations =
+        options.Prune
+          ? Prune.GetLiveDeclarations(options, p, allBlocks.ToList()).ToList()
+          : p.TopLevelDeclarations;
+
+      generator.Line("-- Type constructors");
+      // Always include all type constructors
+      p.TopLevelDeclarations.OfType<TypeCtorDecl>().ForEach(tcd => generator.Visit(tcd));
+      generator.NL();
+
+      generator.Line("-- Type synonyms");
+      liveDeclarations.OfType<TypeSynonymDecl>().ForEach(tcd => generator.Visit(tcd));
```

Comment:
Is it good to mix up the order of `liveDeclarations`? I think a more 1-to-1 translation, possibly even keeping comments, would be nicer.

Reply:
At the moment, doing it this way makes for easier debugging the translation, in my experience so far. Once the approach is more stable, I think preserving more of the original structure would be good (though there may be some cases where Lean could require a slightly different order).